### PR TITLE
Improve bounds in Iterator API

### LINKED ISSUE #23587: IteratorExt::rev should be bounded by DoubleEndedIterator
```rust
fn main() {
    (0..10).step_by(2).rev().next();
}
```
produces the following error:
```
error: type `core::iter::Rev<core::iter::StepBy<_, core::ops::Range<_>>>` does not implement any method in scope named `next`
```

According to @eddyb, `IteratorExt::rev` should be bounded by `DoubleEndedIterator` to prevent this.

EDIT: Updated the example to show the correct error.

### PR DESCRIPTION
**Pull request by aturon**
This commit changes `Iterator`'s API by:
- Generalizing bounds from `Iterator` to `IntoIterator` whenever
  possible, matching the semantics and ergonomics of `for` loops.
- Tightens up a few method-level bounds so that you get an error
  earlier. For example, `rev` did not require `DoubleEndedIterator` even
  though the result is only an `Iterator` when the original iterator was
  double-ended.

Closes #23587

The bound-tightening is technically a:

[breaking-change]

but no code should break in practice.

r? @alexcrichton 
cc @Kimundi

### CODE REVIEW

**Comment by alexcrichton** [comment]
@bors: r+ ac6517f

I'm super stoked for this!

File: src/libcore/iter.rs [hunk 1]
```diff
@@ -207,8 +207,10 @@ pub trait Iterator {
     /// both produce the same output.
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn zip<U: Iterator>(self, other: U) -> Zip<Self, U> where Self: Sized {
-        Zip{a: self, b: other}
+    fn zip<U: Iterator>(self, other: U) -> Zip<Self, U::IntoIter> where
```
Comment (oli-obk) [inline src/libcore/iter.rs]: is the `Iterator` bound on `U` still necessary?

**Comment by Kimundi** [comment]
Looks good at first glance

File: src/libcore/iter.rs (hunk 1 above)
Comment (aturon) [inline src/libcore/iter.rs]: Good catch, will fix.

**Pull request merged**